src/Compilers/CSharp/Portable/Utilities/ValueSetFactory.CharTC.cs

```diff
@@ -47,15 +49,25 @@ bool NumericTC<char>.Related(BinaryOperatorKind relation, char left, char right)
                 }
             }
 
-            char NumericTC<char>.Next(char value)
+            char INumericTC<char>.Next(char value)
             {
                 Debug.Assert(value != char.MaxValue);
                 return (char)(value + 1);
             }
 
-            char EqualableValueTC<char>.FromConstantValue(ConstantValue constantValue) => constantValue.CharValue;
+            char INumericTC<char>.FromConstantValue(ConstantValue constantValue) => constantValue.CharValue;
 
-            string NumericTC<char>.ToString(char value) => $"'{value}'";
+            string INumericTC<char>.ToString(char c)
+            {
+                // The set of Unicode character categories containing non-rendering,
+                // unknown, or incomplete characters.
+                var nonRenderingCategories = new UnicodeCategory[] {
+                    UnicodeCategory.Control,
+                    UnicodeCategory.OtherNotAssigned,
+                    UnicodeCategory.Surrogate };
+                var isPrintable = char.IsWhiteSpace(c) || !nonRenderingCategories.Contains(char.GetUnicodeCategory(c));
```

Comment:
>nonRenderingCategories.Contains(char.GetUnicodeCategory(c) [](start = 59, length = 58)

Consider using `char.GetUnicodeCategory(c) switch ...` rather than constructing an array.